sources/SilkTouch/SilkTouch/Mods/TransformCOM.cs

```diff
@@ -0,0 +1,1027 @@
+            var firstPass = new TypeDiscoverer(
+                cfg.BaseTypes
+                    ?? new() { { "IUnknown.Interface", "Silk.NET.Windows.IUnknown.Interface" } },
+                logger
+            );
+
+            foreach (var comType in cfg.AdditionalCOMTypes ?? [])
+            {
+                firstPass.AddInitialType(
+                    comType.Key,
+                    comType.GetValue<string>("Namespace") ?? "Silk.NET",
+                    comType.GetValue<string>("Parent") ?? string.Empty
+                );
+            }
+
+            foreach (var docId in proj?.DocumentIds ?? [])
+            {
+                index++;
+                var doc =
+                    proj?.GetDocument(docId)
+                    ?? throw new InvalidOperationException("Document missing");
+                if (await doc.GetSyntaxRootAsync(ct) is not { } root)
+                {
+                    continue;
+                }
+
+                firstPass.Visit(root);
+                progressService.SetProgress((float)index / count);
+            }
+
+            //Cleanup our lists
+            firstPass.FoundCOMTypes = firstPass
+                .FoundCOMTypes.Where(val => val.Value.Item1)
+                .ToDictionary();
+
+            index = 0;
+            var rewriter = new Rewriter(firstPass.FoundCOMTypes);
+
+            progressService.SetTask("COM Object Rewrite");
+            foreach (var docId in proj?.DocumentIds ?? [])
+            {
+                index++;
+                var doc =
+                    proj?.GetDocument(docId)
+                    ?? throw new InvalidOperationException("Document missing");
+                if (await doc.GetSyntaxRootAsync(ct) is not { } root)
+                {
+                    continue;
+                }
+
+                doc = doc.WithSyntaxRoot(rewriter.Visit(root).NormalizeWhitespace());
+
+                proj = doc.Project;
+
+                logger.LogDebug("COM Rewrite for {0} Complete ({1}/{2})", doc.Name, index, count);
+                progressService.SetProgress((float)index / count);
+            }
+            ctx.SourceProject = proj;
+        }
+
+        /// <inheritdoc/>
+        public void Transform(
+            MethodDeclarationSyntax current,
+            ITransformationContext ctx,
+            Action<MethodDeclarationSyntax> next
+        )
+        {
+            //pass along the original method
+            next(current);
+
+            if (current.Modifiers.Any(mod => mod.IsKind(SyntaxKind.PublicKeyword)))
+            {
+                int guidPtrIndex = -1;
+                int voidPtrIndex = -1;
+
+                List<ParameterSyntax> paramList = [];
+                List<ArgumentSyntax> argList = [];
+                AssignmentExpressionSyntax? assignment = null;
+
+                //look for guid and void ptr
+                for (var i = 0; i < current.ParameterList.Parameters.Count; i++)
```

Comment:
These should be function transformers, but I think we've had this argument before so ~~I'm happy for~~ I can tolerate this being resolved for now and I'll try and rectify it next time I'm able to work on the generator.

Reply:
This is a function transformer. This is in the transform function and TransformCom is defined as an IFunctionTransformer now.

Reply:
Lol I'm blind, thanks